clientv3/concurrency/election.go

```diff
@@ -0,0 +1,166 @@
+		txn := e.client.Txn(ctx).If(cmp)
+		txn = txn.Then(v3.OpPut(e.leaderKey, val, v3.WithLease(s.Lease())))
+		tresp, terr := txn.Commit()
+		if terr != nil {
+			return terr
+		}
+		if !tresp.Succeeded {
+			e.leaderKey = ""
+			return fmt.Errorf("election: lost leadership key %v", e.leaderKey)
+		}
+		return nil
+	}
+
+	k, rev, err := NewUniqueKV(ctx, e.client, e.keyPrefix, val, v3.WithLease(s.Lease()))
+	if err == nil {
+		err = e.waitLeadership(ctx, k, rev)
+	}
+
+	if err != nil {
+		// clean up in case of context cancel
+		select {
+		case <-ctx.Done():
+			e.client.Delete(e.ctx, k)
+		default:
+		}
+		return err
+	}
+
+	e.leaderKey, e.leaderRev = k, rev
+	return nil
+}
+
+// Resign lets a leader start a new election.
+func (e *Election) Resign() (err error) {
+	if e.leaderKey != "" {
+		_, err = e.client.Delete(e.ctx, e.leaderKey)
+		e.leaderKey = ""
+	}
+	return err
+}
+
+// Leader returns the leader value for the current election.
+func (e *Election) Leader() (string, error) {
+	resp, err := e.client.Get(e.ctx, e.keyPrefix, v3.WithFirstCreate()...)
+	if err != nil {
+		return "", err
+	} else if len(resp.Kvs) == 0 {
+		// no leader currently elected
+		return "", etcdserver.ErrNoLeader
+	}
+	return string(resp.Kvs[0].Value), nil
+}
+
+// Observe returns a channel that observes all elected proposals.
+func (e *Election) Observe(ctx context.Context) <-chan v3.GetResponse {
+	retc := make(chan v3.GetResponse)
+	go e.observe(ctx, retc)
+	return retc
+}
+
+func (e *Election) observe(ctx context.Context, ch chan<- v3.GetResponse) {
+	defer close(ch)
+
+	checkNextLeader := true
+	for checkNextLeader {
+		checkNextLeader = false
+		resp, err := e.client.Get(ctx, e.keyPrefix, v3.WithFirstCreate()...)
+		if err != nil {
+			return
+		}
+
+		var kv *storagepb.KeyValue
+
+		cctx, cancel := context.WithCancel(ctx)
+		if len(resp.Kvs) == 0 {
+			// wait for first key on prefix
+			opts := []v3.OpOption{v3.WithRev(resp.Header.Revision), v3.WithPrefix()}
+			wr, ok := <-e.client.Watch(cctx, e.keyPrefix, opts...)
+			if !ok || len(wr.Events) == 0 {
+				cancel()
+				return
+			}
+			kv = wr.Events[0].Kv
+		} else {
+			kv = resp.Kvs[0]
+		}
+
+		wch := e.client.Watch(cctx, string(kv.Key), v3.WithRev(kv.ModRevision))
+		// get updates until key deletion
+		for wr := range wch {
+			for _, ev := range wr.Events {
+				if ev.Type == storagepb.DELETE {
+					checkNextLeader = true
+					cancel()
+					break
+				}
+				resp.Header = &wr.Header
+				resp.Kvs = []*storagepb.KeyValue{ev.Kv}
+				ch <- *resp
+			}
+		}
+		cancel()
+	}
+}
+
+func (e *Election) waitLeadership(ctx context.Context, tryKey string, tryRev int64) (err error) {
+	for err == nil {
+		var resp *v3.GetResponse
+		resp, err = e.client.Get(ctx, e.keyPrefix, v3.WithFirstCreate()...)
+		if err != nil || string(resp.Kvs[0].Key) == tryKey {
+			break
+		}
+		err = waitUpdate(ctx, e.client, string(resp.Kvs[0].Key), v3.WithRev(tryRev))
```

Comment:
Not necessarily, but it has to tolerate keys dropping out of the election before they attain leadership. The original version was broken in this respect.

I think a better way to do this is to `Get(prefix, WithRev(tryRev-1))` and watch for deletion of all matched keys.